[Extensions] Communicate implemented extension interfaces to OpenSearch

varuntumbe opened the pull request:
Signed-off-by: varuntumbe <varuntumbe1@gmail.com>

<!--  Thanks for sending a pull request, here are some tips:

1. If this is a fix for an undisclosed security vulnerability, please STOP. All security vulnerability reporting and fixes should be done as per our security policy https://github.com/opensearch-project/OpenSearch/security/policy
2. If this is your first time, please read our contributor guidelines: https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md and developer guide https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md
3. Ensure you have added or ran the appropriate tests for your PR: https://github.com/opensearch-project/OpenSearch/blob/main/TESTING.md
-->

### Description
All the work to be done for this PR  ( According to my analysis ):

Work Done :

1) Added some logic in InitializeExtensionResponse.java so that sdk can send implimented Interface along with extension name as a response 
2) Created a data member inside ( List<String> ) DiscoveryExtensionNode.java to hold the implimented Interface values 
3) Added a line in handleResponse method of ExtensionManager by which we can save the implimented interfaces
4) Wrote Testcase for InitializeExtensionResponse.java
Pending :
None

### Issues Resolved
https://github.com/opensearch-project/opensearch-sdk-java/issues/319

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

varuntumbe commented:
Hi @dbwiddis ,

I thought I will open a PR and will add the updates one by one so that mistakes could get corrected faster instead of putting all the changes in the end as PR. Is it okay ?

One Question as of now :

1) Should I writes Request and RequestHandler classes in extensions/settings package ? or should I make a different new package ?

Thanks and Regards
Varun

dbwiddis commented:
> I thought I will open a PR and will add the updates one by one so that mistakes could get corrected faster instead of putting all the changes in the end as PR. Is it okay ?

Totally fine, but please switch the PR to a "Draft" to signal to the project maintainers it's not ready for their review yet.

Also note that while I'm a maintainer on the SDK project, I'm not one here, so I'll do my best to get this ready for review but it will then have to pass muster with OpenSearch maintainers who are sometimes pickier than me ;)

> 1. Should I writes Request and RequestHandler classes in extensions/settings package ? or should I make a different new package ?

Definitely some package in or under `org.opensearch.extensions`.  You can put it there for now, and if we need to move it later, we can.

dbwiddis reviewed (commented):
LGTM so far. 

Personal preference would be to shorten those enum values to avoid the repetition of `EXTENSION_INTERFACE` inside a class called `ExtensionInterfaceType`.   That way later you can just refer to them as `ExtensionInterfaceType.ACTION`.

varuntumbe commented:
Yes makes sense, I will shorten it. Thanks @dbwiddis

varuntumbe commented:
Hi @dbwiddis ,

I have added the rest of the implementation. I have built locally and checked as well. probably forgot to remove print statement which I had used for debugging. Please review it once and inform the  changes which I have to do. I will open a PR in sdk as well.

Thanks and Regards 
Varun

dbwiddis commented on server/src/main/java/org/opensearch/extensions/settings/ImplimentedInterfaceRequestHandler.java:
```diff
@@ -0,0 +1,45 @@
+package org.opensearch.extensions.settings;
```
I don't think we need a whole new registration handler for this (but I may be wrong).

OpenSearch already sends an "ExtensionInitRequest" to the Extension, which currently responds with an `InitializeExtensionResponse` containing just the name.  If we update that response to include this implemented extension enum list/set, that would be a reasonable point to add it to the DiscoveryExtensionNode.

varuntumbe replied:
Yes understood, I will do it that way

Thanks

dbwiddis reviewed (commented):
Suggestions inline

varuntumbe commented on server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:
```diff
@@ -120,13 +123,36 @@ public static enum OpenSearchRequestType {
+        INDEXSTORE,
+        INGEST,
+        MAPPER,
+        PERSISTANTTASK,
```
Noted. Will change

varuntumbe commented on server/src/main/java/org/opensearch/extensions/DiscoveryExtensionNode.java:
```diff
@@ -34,6 +34,7 @@ public class DiscoveryExtensionNode extends DiscoveryNode implements Writeable,
 
     private Version minimumCompatibleVersion;
     private List<ExtensionDependency> dependencies = Collections.emptyList();
+    private List<ExtensionsManager.ExtensionInterfaceType> implimentedInterfaces = Collections.emptyList();
```
Oops sorry, i will change these.

dbwiddis commented on server/src/main/java/org/opensearch/discovery/InitializeExtensionResponse.java:
```diff
@@ -46,18 +49,22 @@
 
     public InitializeExtensionResponse(StreamInput in) throws IOException {
         name = in.readString();
+        this.implementatedInterfaces = new ArrayList<String>(Arrays.asList(in.readStringArray()));
```
Spelling: implemented

Can't you directly assign `Arrays.asList()` without wrapping it in another list constructor?

varuntumbe replied:
Yes, I have done the changes.

dbwiddis commented on server/src/main/java/org/opensearch/discovery/InitializeExtensionResponse.java:
```diff
@@ -68,21 +75,29 @@ public String getName() {
+     */
+
+    public List<String> getImplementedInterfaces() {
+        return implementatedInterfaces;
```
Spelling: implemented

dbwiddis commented on server/src/main/java/org/opensearch/discovery/InitializeExtensionResponse.java:
```diff
@@ -68,21 +75,29 @@ public String getName() {
     @Override
     public String toString() {
-        return "PluginResponse{" + "name" + name + "}";
+        return "PluginResponse{" + "name" + name + "implementedInterfaces" + implementatedInterfaces + "}";
```
need a comma and space after name, before the next string.  Also probably want an `=` after both name and interfaces.

dbwiddis commented on server/src/main/java/org/opensearch/discovery/InitializeExtensionResponse.java:
```diff
@@ -68,21 +75,29 @@ public String getName() {
         if (o == null || getClass() != o.getClass()) return false;
         InitializeExtensionResponse that = (InitializeExtensionResponse) o;
-        return Objects.equals(name, that.name);
+        return Objects.equals(name, that.name) && Objects.equals(implementatedInterfaces, that.implementatedInterfaces);
```
Spelling: implemented

dbwiddis commented on server/src/main/java/org/opensearch/discovery/InitializeExtensionResponse.java:
```diff
@@ -68,21 +75,29 @@ public String getName() {
     @Override
     public int hashCode() {
-        return Objects.hash(name);
+        return Objects.hash(name, implementatedInterfaces);
```
Spelling: implemented

dbwiddis commented on server/src/main/java/org/opensearch/extensions/ExtensionsManager.java:
```diff
@@ -83,6 +83,7 @@ public class ExtensionsManager {
     public static final String REQUEST_EXTENSION_UPDATE_SETTINGS = "internal:discovery/updatesettings";
     public static final String REQUEST_EXTENSION_DEPENDENCY_INFORMATION = "internal:discovery/dependencyinformation";
     public static final String REQUEST_EXTENSION_REGISTER_CUSTOM_SETTINGS = "internal:discovery/registercustomsettings";
+    public static final String REQUEST_EXTENSION_REGISTER_IMPLIMENTED_INTERFACES = "internal:discovery/registerimplimentedinterfaces";
```
Spelling: implemented

varuntumbe replied:
Removing this as this is not needed

dbwiddis commented on server/src/test/java/org/opensearch/discovery/InitializeExtensionResponseTest.java:
```diff
@@ -0,0 +1,51 @@
+public class InitializeExtensionResponseTest extends OpenSearchTestCase {
+    public void testInitialzeExtensionResponse() throws Exception {
+        String expectedName = "testsample-sdk";
+        List<String> expectedImplimentedInterfaces = new ArrayList<>(Arrays.asList("Action", "Search"));
```
Spelling: implemented

dbwiddis reviewed (commented):
Logic LGTM, a few spelling fixes needed.

dbwiddis reviewed (approved):
LGTM!

owaiskazi19 commented on server/src/main/java/org/opensearch/discovery/InitializeExtensionResponse.java:
```diff
@@ -46,18 +48,22 @@
  */
 public class InitializeExtensionResponse extends TransportResponse {
     private String name;
+    private List<String> implementedInterfaces;
```
@dbwiddis need suggestion for a better name instead of calling it `implementedInterfaces`, can we say just `modules`?

dbwiddis replied:
@owaiskazi19 what terminology does the plugin ecosystem use?  

`ActionModule` is a concrete class that is different than `ActionPlugin` interface, so I'm not sure modules is right.

owaiskazi19 replied:
`modules` was just a sample suggestion. Plugin ecosystem filters the plugin based on the class and doesn't have any name for the interfaces. Probably we can call them as `extensionServices` or `extensionInterfaces`?

owaiskazi19 reviewed (commented):
Overall LGTM! Minor nits. Thanks @varuntumbe for your contribution.

owaiskazi19 commented:
Build failure:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':server:testingConventions'.
> Testing conventions [Seem like test classes but don't match naming convention:
    * org.opensearch.discovery.InitializeExtensionResponseTest] are not honored

varuntumbe commented on server/src/test/java/org/opensearch/discovery/InitializeExtensionResponseTest.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class InitializeExtensionResponseTest extends OpenSearchTestCase {
```
Hi @owaiskazi19 ,

Yes, let me push with these changes tomorrow, thanks for the review.

owaiskazi19 commented:
@varuntumbe you need to rebase with the latest `main` to get the successful build. Thanks.

varuntumbe commented:
> @varuntumbe you need to rebase with the latest `main` to get the successful build. Thanks.

Done.

Thanks

dbwiddis commented:
> > @varuntumbe you need to rebase with the latest `main` to get the successful build. Thanks.
> 
> Hi @owaiskazi19, I have rebased and pushed it again. Tell me if there is anything more to do.
> 
> Thanks.

Need to have constructor and filename match:

> /home/runner/work/OpenSearch/OpenSearch/server/src/test/java/org/opensearch/discovery/InitializeExtensionResponseTest.java:20: error: class InitializeExtensionResponseTests is public, should be declared in a file named InitializeExtensionResponseTests.java
> > Task :build-tools:licenseHeaders
> public class InitializeExtensionResponseTests extends OpenSearchTestCase {

varuntumbe commented:
> > > @varuntumbe you need to rebase with the latest `main` to get the successful build. Thanks.
> > 
> > 
> > Hi @owaiskazi19, I have rebased and pushed it again. Tell me if there is anything more to do.
> > Thanks.
> 
> Need to have constructor and filename match:
> 
> > /home/runner/work/OpenSearch/OpenSearch/server/src/test/java/org/opensearch/discovery/InitializeExtensionResponseTest.java:20: error: class InitializeExtensionResponseTests is public, should be declared in a file named InitializeExtensionResponseTests.java
> > > Task :build-tools:licenseHeaders
> > > public class InitializeExtensionResponseTests extends OpenSearchTestCase {


Hi @dbwiddis,

I corrected it. Tell me if there is anything more to change. Happy weekend !

Thanks and Regards
Varun

owaiskazi19 commented on server/src/test/java/org/opensearch/discovery/InitializeExtensionRequestTest.java:
```diff
@@ -0,0 +1,120 @@
+            List.of(expectedDependency)
+        );
+
+        System.out.println(extensionNode.getName());
```
We should not use `System.out` in production code

varuntumbe replied:
Hi @owaiskazi19 ,

Oops sorry, I must have committed this file by mistake. This is not yet complete and testcase errors out. I have written testcase only for InitializeExtensionResponseTest where I have made the change. I will commit this InitializeExtensionRequestTest separately later.

Thanks

varuntumbe commented:
Hi @owaiskazi19,

Any more changes to do ?

Thanks and Regards
Varun.

owaiskazi19 commented:
@varuntumbe we can have the test file`InitializeExtensionRequestTests` as a part of this PR.

varuntumbe commented:
> file`InitializeExtensionRequestTests`

It needs more work and its incomplete currently. might take few days is it okay ? 

Thanks and Regards
Varun

owaiskazi19 commented:
> It needs more work and its incomplete currently. might take few days is it okay ?

You can go ahead and raise another PR for it. Meanwhile, we can get this one in!

ryanbogan commented on server/src/test/java/org/opensearch/discovery/InitializeExtensionResponseTests.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.List;
+
+public class InitializeExtensionResponseTests extends OpenSearchTestCase {
+    public void testInitialzeExtensionResponse() throws Exception {
```
Minor nit: spelling error on initialize

ryanbogan replied:
@varuntumbe I'm also fine fixing this in the upcoming PR that will add the tests

varuntumbe replied:
Hi @ryanbogan ,

Sorry about that. For some reason, I am doing lot of awful spelling mistakes. I corrected this one. Hope everything else is intact !

Thanks and Regards
Varun.

ryanbogan commented:
Waiting to merge until gradle check is finished running, but LGTM!
